### PR TITLE
Dwh298

### PR DESCRIPTION
**Pull request by dwhall**
Issues fixed by this PR
-----------------------
#298

What does this implement/fix?
-----------------------------
Allows setting the name attribute in Button's constructor.

Any other comments?
-------------------
This only fixes half of the issue.  I do not know how to fix the other half (its unit test remains in Error).
This issue may apply to some other widgets, but I'm being lazy and only care about Button right now.

### CODE REVIEW

**Comment by peterbrittain** [comment]
Thanks for this.  First question: why do you want to be able to name your buttons?  They don't change their value when clicked...  I guess this does allow you to find them, though.  Of course you could just save off a reference when constructing them instead.  Anyway...  Tell me more about your use case.

**Review by peterbrittain — CHANGES_REQUESTED** [review]
Assuming you have a good use case, a couple of things need changing...

**Comment by dwhall** [comment]
My use case is that I am trying to create an interactive status bar.  I want to be able to dynamically change the text of the buttons that show information summaries.  Clicking on the buttons activates a Frame to either show the information in greater detail or allow me to edit settings.

File: tests/test_widgets.py [hunk 1]
```diff
@@ -3214,5 +3214,26 @@ def click():
...
+
+    def test_button_name2(self):
+        """
+        Check Button name can be set as an attribute.
+        """
+        def _on_click():
+            pass
+
+        btn = Button("Run", _on_click)
+        btn.name="btn_run"
```
[the remark this answers is not in the document]
Comment (dwhall) [reply tests/test_widgets.py:3235]: Will do.

File: asciimatics/widgets/button.py [hunk 2]
```diff
@@ -25,7 +25,10 @@ def __init__(self, text, on_click, label=None, add_box=True, **kwargs):
 
         Also see the common keyword arguments in :py:obj:`.Widget`.
         """
-        super(Button, self).__init__(None, **kwargs)
+        name = kwargs.get("name", None)
+        if "name" in kwargs:
+            del kwargs["name"]
+        super(Button, self).__init__(name, **kwargs)
```
[the remark this answers is not in the document]
Comment (dwhall) [reply asciimatics/widgets/button.py:31]: Will do.

**Comment by peterbrittain** [comment]
Nice.  Obviously, you don't need names for that as the name isn't what is displayed on the screen.  It sounds like you want a new property to allow you to edit the text.  That said, if you are using the name to find the button object, you'll also need this change to do that.

File: asciimatics/widgets/button.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dwhall) [reply asciimatics/widgets/button.py:31]: I put the 'name' keyword arg at the end of the current keyword args and before '**kwargs' so I don't upset code from those who fill keyword args with positional arguments.  Unfortunately, this puts 'name' out-of-order compared to other widgets's constructors, but I believe not breaking existing code is more important.

**Comment by dwhall** [comment]
I admit I was making an assumption that giving Button() a name would allow me to set its text via the Frame's .data property.  Does this sound like a feature you would want?

File: asciimatics/widgets/button.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (peterbrittain) [reply asciimatics/widgets/button.py:28]: Agreed.  That works for me.

**Comment by peterbrittain** [comment]
They are deliberately not the same thing.  The name is the identifier for the programmer to use when setting/extracting values (and so should never change).  The text is the string to display to the user (and so could change for all sorts of reasons - e.g. using a different language).

What I think you need is a text property.  You could pick the logic from label.py to add the same function for buttons.  Suggest you move the add_box logic to the setter for consistency (we want button.text = "foo" to have the same effect as passing "foo" as the display text to the Button constructor).

**Comment by dwhall** [comment]
Sorry, I meant giving the Button a name so I could use that name in a dict of name:values given to Frame .data to update the text of the Button.

Thanks for the tip, I'll give it a try on my fork.

I don't know github etiquette on who presses "Resolve conversation".  I pressed it trying to see if that got rid of the "1 change requested", but it did not.  Let me know if you need anything else changed.  Otherwise, I'll let you re-review and merge/reject this PR.

File: asciimatics/widgets/button.py [hunk 3]
```diff
@@ -17,16 +17,16 @@ class Button(Widget):
 
     __slots__ = ["_text", "_add_box", "_on_click", "_label"]
 
-    def __init__(self, text, on_click, label=None, add_box=True, **kwargs):
+    def __init__(self, text, on_click, label=None, add_box=True, name=None, **kwargs):
         """
         :param text: The text for the button.
         :param on_click: The function to invoke when the button is clicked.
         :param label: An optional label for the widget.
```
Comment (peterbrittain) [inline asciimatics/widgets/button.py:24]: Looks good.  But can you please add the docstring for the new parameter.  Feel free to borrow the text from other widgets to keep it quick n easy.

**Comment by peterbrittain** [comment]
That's all good once you add the docs.  If you want you can also add yourself to credits.py if you'd like to see your name in lights!

**Review by peterbrittain — CHANGES_REQUESTED** [review]
One other thought...  Could you please add a line to CHANGES to say we've added this new attribute to Button?

File: asciimatics/widgets/button.py (hunk 3 above)
Comment (dwhall) [reply asciimatics/widgets/button.py:24]: Sure thing.  Added a docstring for add_box while I was there.

Comment (peterbrittain) [reply asciimatics/widgets/button.py:24]: Nice!  Thanks

**Comment by peterbrittain** [comment]
Perfect!  Many thanks.

**Pull request merged**